platform/features/table/src/controllers/MCTTableController.js

```diff
@@ -51,22 +61,84 @@ define(
                 self.updateRows($scope.rows);
             };
 
+            /*
+             * Define watches to listen for changes to headers and rows.
+             */
             $scope.$watchCollection('filters', function () {
-                self.updateRows(self.$scope.rows);
+                self.updateRows($scope.rows);
             });
-            $scope.$watchCollection('headers', this.updateHeaders.bind(this));
-            $scope.$watchCollection('rows', this.updateRows.bind(this));
+            $scope.$watch('headers', this.updateHeaders.bind(this));
+            $scope.$watch('rows', this.updateRows.bind(this));
+
+            /*
+             * Listen for rows added individually (eg. for real-time tables)
+             */
+            $scope.$on('addRow', this.newRow.bind(this));
         }
 
         /**
-         * On scroll, calculate which rows indexes are visible and
-         * ensure that an equal number of rows are preloaded for
-         * scrolling in either direction.
+         * If auto-scroll is enabled, this function will scroll to the
+         * bottom of the page
+         * @private
+         */
+        MCTTableController.prototype.scrollToBottom = function() {
+            var self = this;
+
+            //Use timeout to defer execution until next digest when any
+            // pending UI changes have completed, eg. a new row in the table.
+            if (this.$scope.autoScroll) {
+                this.$timeout(function(){
+                        self.scrollable[0].scrollTop = self.scrollable[0].scrollHeight;
+                });
+            }
+        };
+
+        /**
+         * Handles a row add event. Rows can be added as needed using the
+         * `addRow` broadcast event.
+         * @private
+         */
+        MCTTableController.prototype.newRow = function (event, row) {
+            //Add row to the filtered, sorted list of all rows
+            if (this.filterRows([row]).length > 0) {
+                this.insertSorted(this.$scope.displayRows, row);
+            }
+
+            //Keep 'rows' synchronized as it provides the unsorted,
+            // unfiltered model for this view
+            if (!this.$scope.rows) {
+                this.$scope.rows = [];
+            }
+            this.$scope.rows.push(row);
```

Comment:
What about `$scope.broadcast('add:row', index)` -> `addRow($scope.rows[index])`?  What I really want to avoid is the ambiguous case where either object could modify the underlying data.

Let me point out two other performant frameworks with some pseudocode: Backbone and UIKit (iOS).

Backbone: you'd pass an object which the view would observe and update as changes happen.

```javascript
// not full implementation.
TableView.prototype.initialize = function () {
    this.listenTo(this.collection, 'reset', this.resetRows);
    this.listenTo(this.collection, 'add', this.addRow);
}

var table = new TableView({collection: TelemetryCollection, otherOptions...});
$container.append(table.render().$el);
```

UICollectionView (iOS): you'd pass a data source, which has methods for fetching number of items and fetching items at a given index.  (there's a few ways to use it, TBH.)

```objectivec
collectionView = [UICollectionView initWithFrame:someFrame collectionViewLayout:someLayout];

collectionView.dataSource = myDataSource;

// and then you can tell the view to update from the data source.
[collectionView reloadData];
[collectionView insertItemsAtIndexPaths:items];
```